File: packages/php/tests/ClientTest.php
<?php

declare(strict_types=1);

namespace Axene\Mailer\Tests;

use Axene\Mailer\AxeneException;
use Axene\Mailer\Client;
use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Middleware;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;

/**
 * Mock-transport tests. No network: a Guzzle MockHandler returns canned
 * responses and a history middleware captures the outgoing requests so we can
 * assert on the wire shape.
 */
final class ClientTest extends TestCase
{
    /** @var list<array{request: Request}> */
    private array $history = [];

    /**
     * Build a client whose Guzzle transport replays the given queued responses
     * and records every outgoing request into {@see self::$history}.
     *
     * @param list<Response> $responses
     */
    private function clientWith(array $responses): Client
    {
        $this->history = [];
        $mock = new MockHandler($responses);
        $stack = HandlerStack::create($mock);
        $stack->push(Middleware::history($this->history));
        $guzzle = new GuzzleClient(['handler' => $stack]);

        return new Client('axm_k_test123', ['http' => $guzzle]);
    }

    private function lastRequest(): Request
    {
        /** @var Request $request */
        $request = $this->history[array_key_last($this->history)]['request'];

        return $request;
    }

    /**
     * @param array<string, mixed> $payload
     */
    private function jsonResponse(int $status, array|string $payload): Response
    {
        return new Response($status, ['Content-Type' => 'application/json'], json_encode($payload));
    }

    public function testSendsBearerAuthorizationHeader(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(202, ['id' => 'em_1', 'status' => 'queued']),
        ]);

        $client->emails->send([
            'from' => 'hello@yourdomain.com',
            'to' => 'customer@example.com',
            'subject' => 'Hi',
            'html' => '<p>hi</p>',
        ]);

        $this->assertSame('Bearer axm_k_test123', $this->lastRequest()->getHeaderLine('Authorization'));
    }

    public function testMapsFromToFromUnderscoreAndAddressSugar(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(202, ['id' => 'em_1', 'status' => 'queued']),
        ]);

        $client->emails->send([
            'from' => 'hello@yourdomain.com',
            'to' => 'customer@example.com',
            'subject' => 'Hi',
        ]);

        $body = json_decode((string) $this->lastRequest()->getBody(), true);

        $this->assertArrayHasKey('from_', $body);
        $this->assertArrayNotHasKey('from', $body);
        $this->assertSame(['email' => 'hello@yourdomain.com'], $body['from_']);
        // Bare string `to` becomes a list of address objects.
        $this->assertSame([['email' => 'customer@example.com']], $body['to']);
        // Null fields are dropped.
        $this->assertArrayNotHasKey('html', $body);
        $this->assertArrayNotHasKey('cc', $body);
    }

    public function testSendBatchPostsBareArray(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(202, ['total' => 2, 'sent' => 2, 'failed' => 0, 'results' => []]),
        ]);

        $client->emails->sendBatch([
            ['from' => 'a@x.io', 'to' => 'b@x.io', 'subject' => 'One'],
            ['from' => 'a@x.io', 'to' => 'c@x.io', 'subject' => 'Two'],
        ]);

        $request = $this->lastRequest();
        $this->assertSame('/v1/emails/batch', $request->getUri()->getPath());
        $body = json_decode((string) $request->getBody(), true);
        $this->assertTrue(array_is_list($body), 'batch body must be a bare array');
        $this->assertCount(2, $body);
        $this->assertSame(['email' => 'a@x.io'], $body[0]['from_']);
        $this->assertSame('Two', $body[1]['subject']);
    }

    public function testValidateSendsFullBody(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(200, ['valid' => true, 'can_send' => true, 'issues' => []]),
        ]);

        $client->emails->validate([
            'from' => ['email' => 'a@x.io', 'name' => 'A'],
            'to' => [['email' => 'b@x.io'], 'c@x.io'],
            'cc' => 'd@x.io',
            'subject' => 'Full',
            'html' => '<p>h</p>',
            'text' => 't',
            'replyTo' => 'reply@x.io',
            'tags' => ['welcome'],
        ]);

        $request = $this->lastRequest();
        $this->assertSame('/v1/emails/validate', $request->getUri()->getPath());
        $body = json_decode((string) $request->getBody(), true);

        $this->assertSame(['email' => 'a@x.io', 'name' => 'A'], $body['from_']);
        $this->assertSame([['email' => 'b@x.io'], ['email' => 'c@x.io']], $body['to']);
        $this->assertSame([['email' => 'd@x.io']], $body['cc']);
        $this->assertSame(['email' => 'reply@x.io'], $body['reply_to']);
        $this->assertSame('<p>h</p>', $body['html']);
        $this->assertSame('t', $body['text']);
        $this->assertSame(['welcome'], $body['tags']);
    }

    public function testUploadCsvSendsMultipartFileField(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(200, ['imported' => 3, 'skipped' => 0, 'errors' => []]),
        ]);

        $result = $client->contacts->uploadCsv('list_1', "email,name\na@x.io,A\n", 'people.csv');

        $this->assertSame(3, $result['imported']);
        $request = $this->lastRequest();
        $this->assertSame('/v1/contacts/list_1/upload', $request->getUri()->getPath());

        $contentType = $request->getHeaderLine('Content-Type');
        $this->assertStringContainsString('multipart/form-data', $contentType);

        $rawBody = (string) $request->getBody();
        $this->assertStringContainsString('name="file"', $rawBody);
        $this->assertStringContainsString('filename="people.csv"', $rawBody);
        $this->assertStringContainsString('a@x.io,A', $rawBody);
    }

    public function testSuppressionsListParsesEnvelopeAndAddMapsEmailAddress(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(200, [
                'items' => [['id' => 's_1', 'email_address' => 'spam@x.io', 'reason' => 'manual']],
                'total' => 1,
                'page' => 0,
                'limit' => 50,
            ]),
            $this->jsonResponse(201, ['id' => 's_2', 'email_address' => 'bad@x.io', 'reason' => 'manual']),
        ]);

        $page = $client->suppressions->list(['page' => 0, 'limit' => 50]);
        $this->assertSame(1, $page['total']);
        $this->assertSame('spam@x.io', $page['items'][0]['email_address']);

        $client->suppressions->add(['email' => 'bad@x.io']);
        $body = json_decode((string) $this->lastRequest()->getBody(), true);
        $this->assertSame('bad@x.io', $body['email_address']);
        $this->assertArrayNotHasKey('email', $body);
        $this->assertSame('manual', $body['reason']);
    }

    public function testWebhookUpdateMapsIsActive(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(200, ['id' => 'wh_1', 'url' => 'https://x.io/hook', 'is_active' => false]),
        ]);

        $client->webhooks->update('wh_1', ['isActive' => false]);

        $request = $this->lastRequest();
        $this->assertSame('PATCH', $request->getMethod());
        $body = json_decode((string) $request->getBody(), true);
        $this->assertArrayHasKey('is_active', $body);
        $this->assertFalse($body['is_active']);
        $this->assertArrayNotHasKey('isActive', $body);
    }

    public function testRetriesOn429ThenSucceeds(): void
    {
        $client = $this->clientWith([
            new Response(429, ['Retry-After' => '0', 'Content-Type' => 'application/json'], '{"detail":"slow down"}'),
            $this->jsonResponse(200, [['id' => 'em_1', 'status' => 'sent']]),
        ]);

        $emails = $client->emails->list();

        $this->assertCount(1, $emails);
        $this->assertSame('em_1', $emails[0]['id']);
        $this->assertCount(2, $this->history, 'should have retried once');
    }

    public function testMapsErrorEnvelopeToAxeneException(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(422, ['detail' => ['code' => 'unverified_sender', 'message' => 'Sender not verified']]),
        ]);

        try {
            $client->emails->send(['from' => 'a@x.io', 'to' => 'b@x.io', 'subject' => 'Hi']);
            $this->fail('expected AxeneException');
        } catch (AxeneException $e) {
            $this->assertSame(422, $e->getStatus());
            $this->assertSame('unverified_sender', $e->getCode());
            $this->assertSame('Sender not verified', $e->getMessage());
        }
    }

    public function testZeroBasedPageInQueryString(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(200, []),
        ]);

        $client->emails->list(['page' => 0, 'limit' => 20]);

        $query = $this->lastRequest()->getUri()->getQuery();
        $this->assertStringContainsString('page=0', $query);
        $this->assertStringContainsString('limit=20', $query);
    }

    public function testBulkSendInjectsContactListId(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(200, ['queued' => 5, 'skipped' => 0, 'errors' => []]),
        ]);

        $client->contacts->bulkSend('list_42', [
            'senderAddressId' => 'sa_1',
            'subject' => 'Hello {{name}}',
            'html' => '<p>hi</p>',
        ]);

        $body = json_decode((string) $this->lastRequest()->getBody(), true);
        $this->assertSame('list_42', $body['contact_list_id']);
        $this->assertSame('sa_1', $body['sender_address_id']);
        $this->assertArrayNotHasKey('text', $body, 'null text dropped');
    }

    public function testTemplateCreateMapsHtmlAndTextToBodies(): void
    {
        $client = $this->clientWith([
            $this->jsonResponse(201, ['id' => 'tpl_1', 'name' => 'Welcome']),
        ]);

        $client->templates->create([
            'name' => 'Welcome',
            'html' => '<p>{{name}}</p>',
            'text' => 'hi {{name}}',
        ]);

        $body = json_decode((string) $this->lastRequest()->getBody(), true);
        $this->assertSame('<p>{{name}}</p>', $body['html_body']);
        $this->assertSame('hi {{name}}', $body['text_body']);
        $this->assertArrayNotHasKey('html', $body);
        $this->assertArrayNotHasKey('text', $body);
    }
}

File: packages/ruby/test/client_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class ClientTest < Minitest::Test
  def setup
    @server = MockServer.new
    @client = Axene::Mailer::Client.new(api_key: "axm_k_test", base_url: @server.base_url, max_retries: 3)
  end

  def teardown
    @server.shutdown
  end

  def last_request
    @server.requests.last
  end

  def last_body
    JSON.parse(last_request.body, symbolize_names: true)
  end

  def test_bearer_header_is_sent
    @server.enqueue(status: 200, body: [])
    @client.emails.list
    assert_equal "Bearer axm_k_test", last_request.headers["authorization"]
  end

  def test_send_maps_from_to_from_underscore_and_address_sugar
    @server.enqueue(status: 202, body: { id: "em_1", status: "queued" })
    result = @client.emails.send(
      from: "hi@you.io",
      to: "person@example.com",
      subject: "Hello",
      html: "<p>hi</p>"
    )
    body = last_body
    assert_equal({ email: "hi@you.io" }, body[:from_])
    refute body.key?(:from)
    assert_equal [{ email: "person@example.com" }], body[:to]
    assert_equal "em_1", result[:id]
  end

  def test_send_accepts_named_addresses_and_drops_nil_keys
    @server.enqueue(status: 202, body: { id: "em_2", status: "queued" })
    @client.emails.send(
      from: { email: "hi@you.io", name: "You" },
      to: [{ email: "a@example.com", name: "A" }],
      subject: "Hi"
    )
    body = last_body
    assert_equal({ email: "hi@you.io", name: "You" }, body[:from_])
    refute body.key?(:html)
    refute body.key?(:text)
    refute body.key?(:cc)
  end

  def test_send_batch_posts_a_bare_array
    @server.enqueue(status: 202, body: { total: 2, sent: 2, failed: 0, results: [] })
    @client.emails.send_batch([
                                { from: "a@you.io", to: "x@example.com", subject: "1" },
                                { from: "a@you.io", to: "y@example.com", subject: "2" }
                              ])
    body = JSON.parse(last_request.body, symbolize_names: true)
    assert_kind_of Array, body
    assert_equal 2, body.length
    assert_equal({ email: "a@you.io" }, body[0][:from_])
    assert_equal "/v1/emails/batch", last_request.path
  end

  def test_validate_sends_full_send_body
    @server.enqueue(status: 200, body: { valid: true, can_send: true, issues: [] })
    @client.emails.validate(from: "a@you.io", to: "x@example.com", subject: "Check", text: "body")
    body = last_body
    assert_equal "/v1/emails/validate", last_request.path
    assert_equal({ email: "a@you.io" }, body[:from_])
    assert_equal [{ email: "x@example.com" }], body[:to]
    assert_equal "Check", body[:subject]
    assert_equal "body", body[:text]
  end

  def test_list_uses_zero_based_page_query
    @server.enqueue(status: 200, body: [])
    @client.emails.list(status: "delivered", page: 0, limit: 20)
    q = last_request.query
    assert_includes q, "page=0"
    assert_includes q, "status=delivered"
  end

  def test_multipart_csv_upload_uses_field_file
    @server.enqueue(status: 200, body: { imported: 3, skipped: 0, errors: [] })
    result = @client.contacts.upload_csv("list_1", "email\na@example.com\n", filename: "people.csv")
    req = last_request
    assert_equal "/v1/contacts/list_1/upload", req.path
    assert_includes req.headers["content-type"], "multipart/form-data"
    assert_includes req.headers["content-type"], "boundary="
    assert_includes req.body, %(name="file")
    assert_includes req.body, %(filename="people.csv")
    assert_includes req.body, "a@example.com"
    assert_equal 3, result[:imported]
  end

  def test_suppressions_envelope_and_email_address_mapping
    @server.enqueue(status: 200, body: { items: [{ id: "s1", email_address: "x@example.com", reason: "manual" }],
                                         total: 1, page: 0, limit: 50 })
    page = @client.suppressions.list(page: 0, limit: 50)
    assert_equal 1, page[:total]
    assert_equal "x@example.com", page[:items][0][:email_address]

    @server.enqueue(status: 201, body: { id: "s2", email_address: "y@example.com", reason: "manual" })
    @client.suppressions.add(email: "y@example.com")
    body = last_body
    assert_equal "y@example.com", body[:email_address]
    refute body.key?(:email)
    assert_equal "manual", body[:reason]
  end

  def test_webhooks_is_active_mapping
    @server.enqueue(status: 200, body: { id: "wh_1", url: "https://x.io", events: [], is_active: false })
    @client.webhooks.update("wh_1", is_active: false)
    body = last_body
    assert_equal false, body[:is_active]
    refute body.key?(:isActive)
  end

  def test_templates_html_text_map_to_body_fields
    @server.enqueue(status: 201, body: { id: "tpl_1", name: "Welcome" })
    @client.templates.create(name: "Welcome", html: "<p>hi</p>", text: "hi")
    body = last_body
    assert_equal "<p>hi</p>", body[:html_body]
    assert_equal "hi", body[:text_body]
    refute body.key?(:html)
    refute body.key?(:text)
  end

  def test_contacts_bulk_send_injects_contact_list_id
    @server.enqueue(status: 200, body: { queued: 5, skipped: 0, errors: [] })
    @client.contacts.bulk_send("list_9", sender_address_id: "sa_1", subject: "Hi")
    body = last_body
    assert_equal "list_9", body[:contact_list_id]
    assert_equal "sa_1", body[:sender_address_id]
  end

  def test_retry_then_success_on_429
    @server.enqueue(status: 429, body: { detail: "slow down" }, headers: { "Retry-After" => "0" })
    @server.enqueue(status: 200, body: [])
    result = @client.emails.list
    assert_equal [], result
    assert_equal 2, @server.requests.length
  end

  def test_error_mapping_from_detail_object
    @server.enqueue(status: 400, body: { detail: { code: "invalid", message: "bad input" } })
    err = assert_raises(Axene::Mailer::Error) { @client.emails.list }
    assert_equal 400, err.status
    assert_equal "invalid", err.code
    assert_equal "bad input", err.message
  end
end
